DataDog/java-dogstatsd-client
# [CONTINT-1132] Send in-<inode> using the cgroup controller when container-id cannot be retrieved

Implementation of https://github.com/DataDog/datadog-go/pull/291 but in Java

This Pull Request implements the support of Origin Detection when the container-id is unavailable from inside the application container only with cgroupv2.

It first retrieves the cgroup node path by parsing proc/self/cgroup which is formatted differently depending on the cgroup version. See https://man7.org/linux/man-pages/man7/cgroups.7.html.

The format is a list of hierarchy-ID:controller-list:cgroup-path  similar to:

0::<path>
1:name=systemd:... // only in cgroupv1
Once the path is retrieved, we retrieve the inode of /sys/fs/cgroup + <controller> + <path> where <path> is the path retrieved previously.

I am not sure how to create a dummy app that uses this library.

## Test

- Set up a cgroupv2 environment. A kind cluster should be enough.
- Deploy the agent 7.51.0+ with origin detection and dogstatsd enabled. Using helm:
```
  dogstatsd:
    port: 8125
    originDetection: true
    useHostPort: true
    useHostPID: true
    tagCardinality: high
```
- Note that you will also need the env var:
```
    - name: "DD_DOGSTATSD_ORIGIN_DETECTION_CLIENT"
      value: "true"
```
- Deploy a dummy app. I already made one with and without this change:
```
apiVersion: apps/v1
kind: Deployment
metadata:
  name: my-java-dsd-app
spec:
  replicas: 1
  selector:
    matchLabels:
      app: my-java-dsd-app
  template:
    metadata:
      labels:
        app: my-java-dsd-app
        admission.datadoghq.com/enabled: "false"
    spec:
      containers:
      - name: my-java-dsd-app
        image: docker.io/alidatadog/dummy-java-dsd-app:with-inode@sha256:a62e653e3c8e493611ba5841d3af94921edef0670776d0555fed9eecc2b50341 # use the tag without-inode for comparison
        imagePullPolicy: Always
        env:
        - name: DD_AGENT_HOST
          valueFrom:
            fieldRef:
              fieldPath: status.hostIP
```
- Verify that `dummy.metric` is correctly tagged with container-id when it wasn't before.
If you are curious about the dummy app, I pushed it here https://github.com/AliDatadog/dummy-dsd-java-app.
